src/vault/matrix/storage.rs

```diff
@@ -0,0 +1,596 @@
+        let request_url = format!(
+            "/_matrix/client/r0/user/@{username}/account_data/m.secret_storage.default_key",
+            username = self.username
+        );
+
+        let key_id = self.make_get_request::<KeyID>(&request_url).await?;
+
+        // get https://{server}/_matrix/client/r0/user/{username}/account_data/m.secret_storage.key.{id_from_previous_request}
+        let request_url = format!(
+            "/_matrix/client/r0/user/@{username}/account_data/m.secret_storage.key.{id_from_previous_request}",
+            username = self.username, id_from_previous_request= key_id.key
+        );
+
+        self.make_get_request::<KeyData>(&request_url).await
+    }
+
+    async fn make_get_request<'a, T>(&self, url: &str) -> Result<T>
+    where
+        T: DeserializeOwned,
+    {
+        let res: T = self
+            .client
+            .get(url)
+            .header("Authorization", format!("Bearer {}", &self.token))
+            .recv_json()
+            .await
+            .map_err(|err| {
+                MatrixError::ConnectionFailed
+            })?;
+
+        Ok(res)
+    }
+
+
+    // async fn decode_and_validate_storage_key(&self, storage_key: &str) -> Result<MatrixStorageKey> {
+    //     // decode key
+    //     let mut decoded_key = MatrixKeyOps::decode_recovery_key(storage_key)?;
+    //     // validate key (retrieve key_data)
+    //     decoded_key.key_data = self.get_storage_key_data().await.map(|a| Some(a))?;
+    //     MatrixKeyOps::validate_key(&decoded_key)?;
+    //     Ok(decoded_key)
+    // }
+
+    //  SaveSecret
+
+    async fn save_secret_using_keyfile(
+        &self,
+        secret_name: &str,
+        secret: &[u8],
+        storage_key: &str,
+    ) -> Result<()> {
+        // decode key
+        let mut decoded_key = MatrixKeyOps::decode_recovery_key(storage_key)?;
+        // validate key (retrieve key_data)
+        decoded_key.key_data = self.get_storage_key_data().await.map(|a| Some(a))?;
+        MatrixKeyOps::validate_key(&decoded_key)?;
+
+        // 4: generar seed o recibirla del user
+        // 5: encriptar seed con la llave generada de passphrase/keyfile
+        let encrypted = MatrixKeyOps::encrypt_bytes(
+            secret,
+            &MatrixKeyOps::derive_keys(&decoded_key)?,
+            &MatrixKeyOps::create_iv(),
+        )?;
+
+        print!("Ecrypted secret {:?}", encrypted);
+
+        let secret = MatrixStorageSecret { encrypted };
+        // 6: guardar seed y datos para desencriptar la llave
+        self.save_secret(secret_name, &secret).await
+    }
+
+    async fn save_secret_using_passphrase(
+        &self,
+        secret_name: &str,
+        secret: &[u8],
+        passphrase: &str,
+    ) -> Result<()> {
+        let key_data = self.get_storage_key_data().await?;
+        let mut storage_key = MatrixKeyOps::get_key_from_passphrase(passphrase, &key_data)?;
+
+        storage_key.key_data = self.get_storage_key_data().await.map(|a| Some(a))?;
```

Comment:
I don't think this second call to `get_storage_key_data` is needed specially if we can change the `from_passphrase` to take ownership of `key_data` and assign it already to the `storage_key`